internal/app/connectconformance/server_runner.go

```diff
@@ -0,0 +1,171 @@
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	conformancev1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	"google.golang.org/protobuf/proto"
+)
+
+// runTestCasesForServer runs starts a server process and runs the given test cases while
+// it is active. The test cases are executed by serializing the request, writing to the
+// given requestWriter, and then awaiting a corresponding response to be read from the
+// given responseReader. The given results are used to record the state of all cases and
+// their actual responses.
+//
+// The given cancel function should be invoked if an I/O error occurs while interacting
+// with the requestWriter or responseReader.
+//
+// If isReferenceServer is true, then the server's stderr will be examined as well, to
+// record out-of-band feedback about the client requests.
+func runTestCasesForServer(
+	ctx context.Context,
+	isReferenceServer bool,
+	meta serverInstance,
+	testCases []*conformancev1alpha1.TestCase,
+	startServer processStarter,
+	results *testResults,
+	client clientRunner,
+) {
+	expectations := make(map[string]*conformancev1alpha1.ClientResponseResult, len(testCases))
+	for _, testCase := range testCases {
+		expectations[testCase.Request.TestName] = testCase.ExpectedResponse
+	}
+
+	procCtx, procCancel := context.WithCancel(ctx)
+	defer procCancel()
+	serverProcess, err := startServer(procCtx, isReferenceServer)
+	if err != nil {
+		results.failedToStart(testCases, fmt.Errorf("error starting server: %w", err))
+		return
+	}
+	defer serverProcess.abort()
+	serverProcess.whenDone(func(_ error) {
+		procCancel()
+	})
+
+	var refServerFinished chan struct{}
+	if isReferenceServer { //nolint:nestif
+		refServerFinished = make(chan struct{})
+		go func() {
+			defer close(refServerFinished)
+			r := bufio.NewReader(serverProcess.stderr)
+			for {
+				str, err := r.ReadString('\n')
+				str = strings.TrimSpace(str)
+				if str != "" {
+					parts := strings.SplitN(str, ": ", 2)
+					if len(parts) == 2 {
+						if _, ok := expectations[parts[0]]; ok {
+							// appears to be valid message in the form "test case: error message"
+							results.recordServerSideband(parts[0], parts[1])
+						}
+					}
+				}
+				if err != nil {
+					return
+				}
+			}
+		}()
+	}
+
+	// Write server request.
+	serverRequestData, err := proto.Marshal(&conformancev1alpha1.ServerCompatRequest{
+		Protocol:    meta.protocol,
+		HttpVersion: meta.httpVersion,
+		UseTls:      meta.useTLS,
+	})
+	if err != nil {
+		results.failedToStart(testCases, fmt.Errorf("error writing server request: %w", err))
+		return
+	}
+	if _, err := serverProcess.stdin.Write(serverRequestData); err != nil {
+		results.failedToStart(testCases, fmt.Errorf("error writing server request: %w", err))
+		return
+	}
+	if err := serverProcess.stdin.Close(); err != nil {
+		results.failedToStart(testCases, fmt.Errorf("error writing server request: %w", err))
```

Comment:
I figured this is _highly unlikely_ to ever happen in practice and, if it did, that the actual error value should hopefully describe the problem adequately. I think confusion is unlikely since any write or close errors will be I/O errors and marshaling will be something about incorrect the message/data being marshaled.